pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,417 @@
+
+				stopChan = make(chan bool)
+
+				go c.replayLoop(ctx, gvksToRelistForLoop, stopChan)
+				go gvksFailingToListReconciler(stopChan)
+			}
+			c.mu.Unlock()
+		}
+	}
+}
+
+func (c *CacheManager) replayLoop(ctx context.Context, gvksToRelist []schema.GroupVersionKind, stopChan <-chan bool) {
+	for _, gvk := range gvksToRelist {
+		select {
+		case <-ctx.Done():
+			return
+		case <-stopChan:
+			return
+		default:
+			backoff := wait.Backoff{
+				Duration: time.Second,
+				Factor:   2,
+				Jitter:   0.1,
+				Steps:    3,
+			}
+
+			operation := func() (bool, error) {
+				if err := c.listAndSyncDataForGVK(ctx, gvk); err != nil {
+					return false, err
+				}
+
+				return true, nil
+			}
+
+			if err := wait.ExponentialBackoff(backoff, operation); err != nil {
+				log.Error(err, "internal: error listings gvk cache data", "gvk", gvk)
+			}
+		}
+	}
+}
+
+// listAndSyncData returns a set of gvks that were successfully listed and synced.
+func (c *CacheManager) listAndSyncData(ctx context.Context, gvks []schema.GroupVersionKind) *watch.Set {
+	gvksSuccessfullySynced := watch.NewSet()
+	for _, gvk := range gvks {
+		err := c.listAndSyncDataForGVK(ctx, gvk)
+		if err != nil {
+			log.Error(err, "internal: error syncing gvks cache data")
+			// we don't remove this gvk as we will try to re-add it later
+			// we also don't return on this error to be able to list and sync
+			// other gvks in order to protect against a bad gvk.
+		} else {
+			gvksSuccessfullySynced.Add(gvk)
+		}
+	}
+	return gvksSuccessfullySynced
+}
+
+// makeUpdates performs a conditional wipe followed by a replay if necessary as
+// given by the current spec (currentGVKsInAgg, excluderChanged) at the time of the call.
+func (c *CacheManager) makeUpdates(ctx context.Context) {
+	// assumes the caller has lock
+
+	currentGVKsInAgg := watch.NewSet()
+	currentGVKsInAgg.Add(c.gvkAggregator.ListAllGVKs()...)
+
+	if c.watchedSet.Equals(currentGVKsInAgg) && !c.excluderChanged {
+		// nothing to do if both sets are the same and the excluder didn't change
+		// and there are no gvks that need relisting from a previous wipe
+		return
+	}
+
+	gvksToDelete := c.watchedSet.Difference(currentGVKsInAgg)
+	newGVKsToSync := currentGVKsInAgg.Difference(c.watchedSet)
+	gvksToReplay := c.watchedSet.Intersection(currentGVKsInAgg)
+
+	if gvksToDelete.Size() != 0 || newGVKsToSync.Size() != 0 {
+		// in this case we need to replace the watch set again since there
+		// is drift between the aggregator and the currently watched gvks
+		if err := c.replaceWatchSet(ctx, currentGVKsInAgg); err != nil {
+			log.Error(err, "internal: error replacing watch set")
+		}
+	}
+
+	// remove any gvks not needing to be synced anymore
+	// or re evaluate all if the excluder changed.
+	if gvksToDelete.Size() > 0 || c.excluderChanged {
+		if err := c.wipeData(ctx); err != nil {
+			log.Error(err, "internal: error wiping cache")
+		} else {
+			c.excluderChanged = false
+		}
+		c.gvksToRelist.AddSet(gvksToReplay)
```

Comment:
if we just wiped the cache, then we need to replay every GVK that is in the watchSet (which should be the same as every GVK that's in the GVKAggregator).